cirq/protocols/json.py

```diff
@@ -0,0 +1,309 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+from typing import Union, Any, Dict, Optional, List, Callable, Type
+
+import numpy as np
+from typing_extensions import Protocol
+
+from cirq.type_workarounds import NotImplementedType
+
+
+# Note for developers:
```

Comment:
We should probably have a test that checks that every public class is serializable.

Reply:
Sketched out an implementation of this in the latest commit. There are some classes that we probably don't want or need to make serializable and there are some that are lower priority that I don't want to block this PR from being merged

Reply:
one problem with the existing implementation is that it leaves out singletons whose class is private

Reply:
I reworked the logic to try anything that's not a function or module, with an explicit blacklist